# --------------------------------------- Training -------------------------------------

* You can check the comments to see what each cell does. First, read the comments, and then run the cells to see the outputs.


* There are a few exercises in the notebook. The solution of some of them is given. But please first try to complete the exercise yourself and then look at the solution.


* After running all cells and completing all exercises, complete the tasks at the end of the notebook.

## Importing necessary Libraries  

In [1]:
import pandas as pd
import numpy as np

## Importing a dataset for this workshop

## Importing the dataset directly from the source

In [2]:
# Installing a package which is needed to download the dataset from its online source. This package is recommended by the online source at the following
# URL : https://archive.ics.uci.edu/dataset/2/adult. You need to reinstall the kernel after the package has been installed
!pip3 install ucimlrepo

In [3]:
# Downloading the dataset from the online source. The first two lines are given by the online source mentioned above
from ucimlrepo import fetch_ucirepo
# fetch dataset
adult = fetch_ucirepo(id=2)

# Putting data in a pandas dataframe
X = adult.data.features
y = adult.data.targets
data=pd.concat([X,y],axis=1)


In [4]:
#printing data
data

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48837,39,Private,215419,Bachelors,13,Divorced,Prof-specialty,Not-in-family,White,Female,0,0,36,United-States,<=50K.
48838,64,NaN,321403,HS-grad,9,Widowed,NaN,Other-relative,Black,Male,0,0,40,United-States,<=50K.
48839,38,Private,374983,Bachelors,13,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K.
48840,44,Private,83891,Bachelors,13,Divorced,Adm-clerical,Own-child,Asian-Pac-Islander,Male,5455,0,40,United-States,<=50K.


## Importing the dataset from a folder on your local disk

In [ ]:
#importing the dataset as a Pandas DataFrame into Python if the dataset is stored on your local hard disk
# You can download the dataset from the following URL: https://archive.ics.uci.edu/dataset/2/adult
data = pd.read_csv('adult.csv') # Replace the current data path with the data path on your local disk

## Exploring the dataset

In [5]:
# Showing the first 5 rows of the dataset
data.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [6]:
# Finding the shape of the data
print(data.shape)

(48842, 15)


In [7]:
# Generate a dataset by randomly extracting 30000 rows (samples)
data_new = data.sample(n=30000, random_state = 48)

In [8]:
# Printing the new dataset
data_new

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
8029,29,Private,216481,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,>50K
45203,36,Private,280570,Some-college,10,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,45,United-States,<=50K.
27498,25,?,100903,Bachelors,13,Married-civ-spouse,?,Wife,White,Female,0,0,25,United-States,<=50K
48416,47,Private,145636,Assoc-voc,11,Married-civ-spouse,Handlers-cleaners,Husband,White,Male,0,0,48,United-States,>50K.
43230,33,Private,119422,HS-grad,9,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,40,United-States,<=50K.
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37544,20,Private,166371,HS-grad,9,Never-married,Craft-repair,Other-relative,White,Male,0,0,40,United-States,<=50K.
20482,80,Private,202483,HS-grad,9,Married-spouse-absent,Adm-clerical,Not-in-family,White,Female,0,0,16,United-States,<=50K
39667,20,Private,175808,HS-grad,9,Never-married,Craft-repair,Own-child,White,Male,0,0,40,United-States,<=50K.
40001,25,State-gov,31350,Some-college,10,Never-married,Other-service,Not-in-family,White,Male,0,0,40,United-States,<=50K.


In [9]:
# The indices of different rows in the dataset are currently messy. This happens in many data science projects. Always reindex
# the dataset if you are unsure the indices are correct.
data_new.reset_index(drop=True, inplace=True)

In [10]:
# Checking if the indices are correct
data_new

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,29,Private,216481,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,>50K
1,36,Private,280570,Some-college,10,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,45,United-States,<=50K.
2,25,?,100903,Bachelors,13,Married-civ-spouse,?,Wife,White,Female,0,0,25,United-States,<=50K
3,47,Private,145636,Assoc-voc,11,Married-civ-spouse,Handlers-cleaners,Husband,White,Male,0,0,48,United-States,>50K.
4,33,Private,119422,HS-grad,9,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,40,United-States,<=50K.
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29995,20,Private,166371,HS-grad,9,Never-married,Craft-repair,Other-relative,White,Male,0,0,40,United-States,<=50K.
29996,80,Private,202483,HS-grad,9,Married-spouse-absent,Adm-clerical,Not-in-family,White,Female,0,0,16,United-States,<=50K
29997,20,Private,175808,HS-grad,9,Never-married,Craft-repair,Own-child,White,Male,0,0,40,United-States,<=50K.
29998,25,State-gov,31350,Some-college,10,Never-married,Other-service,Not-in-family,White,Male,0,0,40,United-States,<=50K.


In [11]:
data

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48837,39,Private,215419,Bachelors,13,Divorced,Prof-specialty,Not-in-family,White,Female,0,0,36,United-States,<=50K.
48838,64,NaN,321403,HS-grad,9,Widowed,NaN,Other-relative,Black,Male,0,0,40,United-States,<=50K.
48839,38,Private,374983,Bachelors,13,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K.
48840,44,Private,83891,Bachelors,13,Divorced,Adm-clerical,Own-child,Asian-Pac-Islander,Male,5455,0,40,United-States,<=50K.


In [12]:
# Getting statistical information of the dataset for different columns (features)
data.describe(include="all")

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
count,48842.000000,47879,4.884200e+04,48842,48842.000000,48842,47876,48842,48842,48842,48842.000000,48842.000000,48842.000000,48568,48842
unique,NaN,9,NaN,16,NaN,7,15,6,5,2,NaN,NaN,NaN,42,4
top,NaN,Private,NaN,HS-grad,NaN,Married-civ-spouse,Prof-specialty,Husband,White,Male,NaN,NaN,NaN,United-States,<=50K
freq,NaN,33906,NaN,15784,NaN,22379,6172,19716,41762,32650,NaN,NaN,NaN,43832,24720
mean,38.643585,NaN,1.896641e+05,NaN,10.078089,NaN,NaN,NaN,NaN,NaN,1079.067626,87.502314,40.422382,NaN,NaN
std,13.710510,NaN,1.056040e+05,NaN,2.570973,NaN,NaN,NaN,NaN,NaN,7452.019058,403.004552,12.391444,NaN,NaN
min,17.000000,NaN,1.228500e+04,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,1.000000,NaN,NaN
25%,28.000000,NaN,1.175505e+05,NaN,9.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
50%,37.000000,NaN,1.781445e+05,NaN,10.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
75%,48.000000,NaN,2.376420e+05,NaN,12.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,45.000000,NaN,NaN


In [13]:
# Showing the dataset information
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  int64 
 1   workclass       47879 non-null  object
 2   fnlwgt          48842 non-null  int64 
 3   education       48842 non-null  object
 4   education-num   48842 non-null  int64 
 5   marital-status  48842 non-null  object
 6   occupation      47876 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital-gain    48842 non-null  int64 
 11  capital-loss    48842 non-null  int64 
 12  hours-per-week  48842 non-null  int64 
 13  native-country  48568 non-null  object
 14  income          48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [14]:
# Getting the count of different values in the column "education-num"
data['education-num'].value_counts()

,count
education-num,
9,15784
10,10878
13,8025
14,2657
11,2061
7,1812
12,1601
6,1389
4,955


In [15]:
# Getting the count of different values in the column "education"
data['education'].value_counts()

,count
education,
HS-grad,15784
Some-college,10878
Bachelors,8025
Masters,2657
Assoc-voc,2061
11th,1812
Assoc-acdm,1601
10th,1389
7th-8th,955


In [16]:
# Dropping a column
data = data.drop(['fnlwgt'], axis=1)

In [17]:
data.shape

(48842, 14)

In [18]:
# Getting the number of unique values of a column
data['education'].nunique()

16

In [19]:
# Finding how many rows are related to either gender
data['sex'].value_counts()

,count
sex,
Male,32650
Female,16192


In [20]:
# Calculating the average age of different genders in the dataset
data['age'].groupby([data['sex']]).mean()


,age
sex,
Female,36.927989
Male,39.494395


In [21]:
# Getting the average age of different genders in the dataset broken down based on their education
data['age'].groupby([data['sex'],data['education']]).mean()

sex     education   
Female  10th            36.474836
        11th            29.963077
        12th            29.469194
        1st-4th         47.098361
        5th-6th         45.362205
        7th-8th         50.577406
        9th             41.272727
        Assoc-acdm      36.508772
        Assoc-voc       38.111717
        Bachelors       35.880904
        Doctorate       45.557522
        HS-grad         38.910339
        Masters         42.665089
        Preschool       44.083333
        Prof-school     39.878788
        Some-college    33.552657
Male    10th            38.603004
        11th            33.360585
        12th            33.544843
        1st-4th         46.489247
        5th-6th         42.994764
        7th-8th         49.174581
        9th             40.750000
        Assoc-acdm      38.593429
        Assoc-voc       38.949510
        Bachelors       40.242610
        Doctorate       47.993763
        HS-grad         39.151773
        Masters         44.632450
        Preschool       41.440678
        Prof-school     46.143875
        Some-college    37.001791
Name: age, dtype: float64

In [22]:
# Getting the maximum age of different races in the dataset
data['age'].groupby([data['race']]).max()

,age
race,
Amer-Indian-Eskimo,82
Asian-Pac-Islander,90
Black,90
Other,77
White,90


* Exercise 1: Write a code to find how much contribution each sex and occupation category made to the capital-gain on average. Apply the code to the dataset and print the result

In [23]:
# Extracting the age and education columns and creating a new DataFrame using these columns
a=data['age']
b=data['education']
new_data=pd.concat([a,b],axis=1)

In [24]:
data

,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48837,39,Private,Bachelors,13,Divorced,Prof-specialty,Not-in-family,White,Female,0,0,36,United-States,<=50K.
48838,64,NaN,HS-grad,9,Widowed,NaN,Other-relative,Black,Male,0,0,40,United-States,<=50K.
48839,38,Private,Bachelors,13,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K.
48840,44,Private,Bachelors,13,Divorced,Adm-clerical,Own-child,Asian-Pac-Islander,Male,5455,0,40,United-States,<=50K.


* Exercise 2: Write a function that receives the dataset and replace Famle with F and Male with M (please try to write it yourself before checking the answer in the next cell)

In [25]:
# exercise 2 solution
def encode_sex(data):
    data.reset_index(drop=True, inplace=True) # reindexing the dataset in case the dataset index is corrupted
    rows=data.shape[0]
    a=data.loc[:,'sex']
    for i in range(rows):
        if a[i]=="Male":
            data.loc[i,"sex"]="M"
        elif a[i]=="Female":
            data.loc[i,"sex"]="F"
    return data

In [26]:
# Copying the data
data_copy=data.copy()

In [27]:
# Applying the encode_sex function to the copied data
data_encoded=encode_sex(data_copy)
data_encoded.head()

,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,M,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,M,0,0,13,United-States,<=50K
2,38,Private,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,M,0,0,40,United-States,<=50K
3,53,Private,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,M,0,0,40,United-States,<=50K
4,28,Private,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,F,0,0,40,Cuba,<=50K


In [28]:
data

,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48837,39,Private,Bachelors,13,Divorced,Prof-specialty,Not-in-family,White,Female,0,0,36,United-States,<=50K.
48838,64,NaN,HS-grad,9,Widowed,NaN,Other-relative,Black,Male,0,0,40,United-States,<=50K.
48839,38,Private,Bachelors,13,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K.
48840,44,Private,Bachelors,13,Divorced,Adm-clerical,Own-child,Asian-Pac-Islander,Male,5455,0,40,United-States,<=50K.


# Tasks

## P 2.1: Complete the following steps (4%):
## 1- Import the dataset from the URL we used in this workshop.
## 2- Generate a new dataset by randomly extracting 10000 samples.
## 3- Drop the 'income' column, reindex the new dataset and then clean it.
## 4- Print the new dataset and use it for the rest of the tasks

In [29]:
############# WRITE YOUR CODE IN THIS CELL (IF APPLICABLE)  ####################
# Step 1: Import necessary libraries and fetch the dataset
from ucimlrepo import fetch_ucirepo
import pandas as pd

# Fetch the Adult dataset from UCI
newadult = fetch_ucirepo(id=2)
# Combine features and target into one DataFrame
adultFrame = pd.concat([newadult.data.features, newadult.data.targets], axis=1)

#print data
adultFrame

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48837,39,Private,215419,Bachelors,13,Divorced,Prof-specialty,Not-in-family,White,Female,0,0,36,United-States,<=50K.
48838,64,NaN,321403,HS-grad,9,Widowed,NaN,Other-relative,Black,Male,0,0,40,United-States,<=50K.
48839,38,Private,374983,Bachelors,13,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K.
48840,44,Private,83891,Bachelors,13,Divorced,Adm-clerical,Own-child,Asian-Pac-Islander,Male,5455,0,40,United-States,<=50K.


In [31]:
# Step 2: Randomly extract 10,000 samples
adultFrame_sample = data.sample(n=10000, random_state=42)

#print data
adultFrame_sample

,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
7762,18,Private,HS-grad,9,Never-married,Other-service,Not-in-family,White,Male,0,0,20,United-States,<=50K
23881,17,Private,12th,8,Never-married,Sales,Own-child,White,Female,0,0,20,United-States,<=50K
30507,25,Local-gov,HS-grad,9,Never-married,Handlers-cleaners,Other-relative,Black,Male,0,0,40,United-States,<=50K
28911,20,Private,Some-college,10,Never-married,Sales,Own-child,White,Female,0,0,30,United-States,<=50K
19484,47,Private,HS-grad,9,Never-married,Machine-op-inspct,Unmarried,Amer-Indian-Eskimo,Male,0,0,56,Puerto-Rico,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19828,79,Private,7th-8th,4,Widowed,Priv-house-serv,Not-in-family,White,Female,2964,0,30,United-States,<=50K
15814,45,State-gov,HS-grad,9,Divorced,Transport-moving,Not-in-family,White,Male,0,0,48,United-States,<=50K
38403,44,Private,9th,5,Never-married,Priv-house-serv,Unmarried,White,Female,0,0,50,Puerto-Rico,<=50K.
33358,44,Private,Assoc-voc,11,Married-civ-spouse,Tech-support,Husband,White,Male,0,0,45,United-States,<=50K.


In [32]:
# Step 3a: Drop the 'income' target column
adultFrame_sample_cleaned = adultFrame_sample.drop(columns=['income'])

In [33]:
# Step 3b: Reindex so row indices are sequential from 0
adultFrame_sample_cleaned.reset_index(drop=True, inplace=True)

In [34]:
# Step 3c: Clean — replace '?' placeholders with NaN, drop incomplete rows, reindex
adultFrame_sample_cleaned.replace('?', pd.NA, inplace=True)
adultFrame_sample_cleaned.dropna(inplace=True)
adultFrame_sample_cleaned.reset_index(drop=True, inplace=True)

print("Before cleaning:", adultFrame_sample.shape)
print("After cleaning :", adultFrame_sample_cleaned.shape)
adultFrame_sample_cleaned

Before cleaning: (10000, 14)
After cleaning : (9241, 13)


,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country
0,18,Private,HS-grad,9,Never-married,Other-service,Not-in-family,White,Male,0,0,20,United-States
1,17,Private,12th,8,Never-married,Sales,Own-child,White,Female,0,0,20,United-States
2,25,Local-gov,HS-grad,9,Never-married,Handlers-cleaners,Other-relative,Black,Male,0,0,40,United-States
3,20,Private,Some-college,10,Never-married,Sales,Own-child,White,Female,0,0,30,United-States
4,47,Private,HS-grad,9,Never-married,Machine-op-inspct,Unmarried,Amer-Indian-Eskimo,Male,0,0,56,Puerto-Rico
...,...,...,...,...,...,...,...,...,...,...,...,...,...
9236,79,Private,7th-8th,4,Widowed,Priv-house-serv,Not-in-family,White,Female,2964,0,30,United-States
9237,45,State-gov,HS-grad,9,Divorced,Transport-moving,Not-in-family,White,Male,0,0,48,United-States
9238,44,Private,9th,5,Never-married,Priv-house-serv,Unmarried,White,Female,0,0,50,Puerto-Rico
9239,44,Private,Assoc-voc,11,Married-civ-spouse,Tech-support,Husband,White,Male,0,0,45,United-States


## P 2.2: Complete the following steps (4%):
## 1- Determine which columns are categorical and which columns are numerical.
## 2- Encode the categorical columns using the correct method.
## 3- Normalise the numerical columns.
## 4- Print the dataset

In [36]:
############# WRITE YOUR CODE IN THIS CELL (IF APPLICABLE)  ####################
from sklearn.preprocessing import LabelEncoder, MinMaxScaler

# Step 1: Identify categorical and numerical columns automatically
categorical_columns = adultFrame_sample_cleaned.select_dtypes(
    include='object'
).columns.tolist()
numerical_columns = adultFrame_sample_cleaned.select_dtypes(
    include=['int64', 'float64']
).columns.tolist()

print("Categorical columns:", categorical_columns)
print("Numerical columns  :", numerical_columns)

Categorical columns: ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
Numerical columns  : ['age', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']


In [41]:
# Step 2: Encode categorical columns using LabelEncoder
def encode_categorical_columns(df):
    label_encoders = {}
    for col in df.select_dtypes(include='object').columns:
        le = LabelEncoder()
        df[col] = le.fit_transform(df[col].astype(str))
        label_encoders[col] = le
    return df, label_encoders

adultFrame_sample_cleaned, encoders = encode_categorical_columns(
    adultFrame_sample_cleaned
)

In [40]:
# Step 3: Normalise numerical columns using Min-Max Scaling
scaler = MinMaxScaler()
adultFrame_sample_cleaned[numerical_columns] = scaler.fit_transform(
    adultFrame_sample_cleaned[numerical_columns]
)

adultFrame_sample_cleaned

,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country
0,0.013699,2,11,0.533333,4,7,1,4,1,0.00000,0.0,0.193878,38
1,0.000000,2,2,0.466667,4,11,3,4,0,0.00000,0.0,0.193878,38
2,0.109589,1,11,0.533333,4,5,2,2,1,0.00000,0.0,0.397959,38
3,0.041096,2,15,0.600000,4,11,3,4,0,0.00000,0.0,0.295918,38
4,0.410959,2,11,0.533333,4,6,4,0,1,0.00000,0.0,0.561224,32
...,...,...,...,...,...,...,...,...,...,...,...,...,...
9236,0.849315,2,5,0.200000,6,8,1,4,0,0.02964,0.0,0.295918,38
9237,0.383562,5,11,0.533333,0,13,1,4,1,0.00000,0.0,0.479592,38
9238,0.369863,2,6,0.266667,4,8,4,4,0,0.00000,0.0,0.500000,32
9239,0.369863,2,8,0.666667,2,12,0,4,1,0.00000,0.0,0.448980,38


The dataset contains missing data encoded in two forms: true NaN (from the CSV parser) and the string '?'. Both must be addressed; replace('?', pd.NA) followed by dropna() handles both in one pass. After cleaning, the sample reduces from 10,000 to 9,241 rows, meaning approximately 7.6% of sampled rows contained at least one missing or '?' value.

## P 2.3: Write a function to split the dataset in half column-wise and swap the first half and the second half. Apply the funtion to the dataset and print the result (3%)

In [42]:
############# WRITE YOUR CODE IN THIS CELL (IF APPLICABLE)  ####################

def swap_column_halves(df):
    num_cols    = len(df.columns)
    mid         = num_cols // 2
    first_half  = df.iloc[:, :mid]
    second_half = df.iloc[:, mid:]
    swapped_df  = pd.concat([second_half, first_half], axis=1)
    return swapped_df

swapped_df = swap_column_halves(adultFrame_sample_cleaned)
print("Original:", list(adultFrame_sample_cleaned.columns))
print("Swapped :", list(swapped_df.columns))
swapped_df.head()


Original: ['age', 'workclass', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country']
Swapped : ['relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'age', 'workclass', 'education', 'education-num', 'marital-status', 'occupation']


,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,age,workclass,education,education-num,marital-status,occupation
0,1,4,1,0.0,0.0,0.193878,38,0.013699,2,11,0.533333,4,7
1,3,4,0,0.0,0.0,0.193878,38,0.000000,2,2,0.466667,4,11
2,2,2,1,0.0,0.0,0.397959,38,0.109589,1,11,0.533333,4,5
3,3,4,0,0.0,0.0,0.295918,38,0.041096,2,15,0.600000,4,11
4,4,0,1,0.0,0.0,0.561224,32,0.410959,2,11,0.533333,4,6


## P 2.4: Write a function that receives two numerical columns' names and compare their values for all rows. If the value of the first column is greater than the second column, the function should produce True, otherwise, it should produce False. The function should append an additional column to the dataset to store the results of the comparison for all rows. Apply the function to the "age" and "hours-per-week" columns in the dataset and print the result (4%).

In [43]:
############# WRITE YOUR CODE IN THIS CELL (IF APPLICABLE)  ####################

def compare_columns(df, col1, col2):
    df['comparison_result'] = df[col1] > df[col2]
    return df

adultFrame_sample_cleaned = compare_columns(
    adultFrame_sample_cleaned, 'age', 'hours-per-week'
)
print(adultFrame_sample_cleaned[['age', 'hours-per-week', 'comparison_result']].head(10))
adultFrame_sample_cleaned.head()


        age  hours-per-week  comparison_result
0  0.013699        0.193878              False
1  0.000000        0.193878              False
2  0.109589        0.397959              False
3  0.041096        0.295918              False
4  0.410959        0.561224              False
5  0.219178        0.397959              False
6  0.287671        0.397959              False
7  0.054795        0.397959              False
8  0.356164        0.397959              False
9  0.520548        0.346939               True


,age,workclass,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,comparison_result
0,0.013699,2,11,0.533333,4,7,1,4,1,0.0,0.0,0.193878,38,False
1,0.000000,2,2,0.466667,4,11,3,4,0,0.0,0.0,0.193878,38,False
2,0.109589,1,11,0.533333,4,5,2,2,1,0.0,0.0,0.397959,38,False
3,0.041096,2,15,0.600000,4,11,3,4,0,0.0,0.0,0.295918,38,False
4,0.410959,2,11,0.533333,4,6,4,0,1,0.0,0.0,0.561224,32,False
